mgr/cephadm: add agent to push metadata to mgr

adk3798 opened the pull request:
Fixes: https://tracker.ceph.com/issues/51004

Signed-off-by: Adam King <adking@redhat.com>

Still needs cleanup and removal of the old exporter but functionality was there during local testing.

Hasn't really accomplished the goal of simplifying the serve loop code-wise because I left in the old
methods of fetching metadata as a fall back when agent is either not running or hasn't sent anything in a long time.

Maybe a more minor issue but because I'm deploying the binary as a required file for the agent the whole binary is
printed to the debug logs when an agent is deployed. Haven't looked into getting around that yet but it definitely
shouldn't be left that way.

<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

adk3798 commented:
not sure how useful these commits are for review since a lot of them modify the same pieces of code. Maybe better I just squash them all now (and properly added which tracker issues are fixed to the commit)?

dsavineau commented on src/cephadm/cephadm:
```diff
@@ -13,6 +13,7 @@ import os
 import platform
 import pwd
 import random
+import requests
```
We need to add python3-requests to the cephadm package dependency list as well

adk3798 replied:
@dsavineau does the latest commit adding the package dependencies look correct?

liewegas replied:
this dependency might be a problem if it isn't part of the base python install...

liewegas replied:
IIUC urllib.request is in the python standard library?  Can we use that instead?  https://docs.python.org/3/library/urllib.request.html#module-urllib.request

dsavineau commented on src/pybind/mgr/requirements.txt:
```diff
@@ -8,4 +8,5 @@ pyOpenSSL
 execnet
 remoto
 Jinja2
+cherrypy
```
Same here, we need to add python3-cherrypy to the ceph-mgr-cephadm package dependency list.

Daniel-Pivonka reviewed (approved):
functioning as expected when tested on my cluster

tchaikov commented:
> not sure how useful these commits are for review since a lot of them modify the same pieces of code. Maybe better I just squash them all now (and properly added which tracker issues are fixed to the commit)?

yeah, better squashing the changes modifying the code introduced in the this very changeset. the cleanup commits in particular. so that the commits are atomic. if you believe that some of them are relatively self-contained, for instance, the commit which add HTTPS support, i think we can have them as separate commits.

adk3798 commented:
squashed nearly all the commits since later commits mostly modified code added in older commits and it would have just made reviewing even more difficult.

adk3798 commented:
jenkins retest this please

tchaikov commented:
@adk3798 better off squashing these two commits, so the change is atomic.

pcuzner commented:
@adk3798 Hi Adam, couple of dumb questions (I only found out about this today, sorry)

- how are you handling changes in the mgr i.e  mgr failover, or mgr redeployed to other hosts? is the intent to use the raw socket mechanism to mgrlistener to deal with this in some way?
- the agent code as https connection with verify=false? Is the cephadm listener always going to be self-signed, or should enabling this feature allow for a user provided cert 
- we're sending data to the mgr, from anywhere - you're using a keyring to validate that only our hosts are allowed to send data to mgr..right?
- the raw socket handling is ipv4 only - is the goal to keep this at ipv4 or support dual stack?
- any thoughts on how this can be extended to be an agent beyond what the exporter already provides? For example, tell the agent to deploy a daemon, so the mgr offloads not just the data gathering but also the deployment work? As it stands the code is heavily focused on sending state, not enabling the offload of actions to help with scale...right?
- how are agents going offline being handled? For example, if 1 agent is not sending data is that a healthcheck...what about 10 hosts?

adk3798 commented:
> @adk3798 Hi Adam, couple of dumb questions (I only found out about this today, sorry)
> 
> * how are you handling changes in the mgr i.e  mgr failover, or mgr redeployed to other hosts? is the intent to use the raw socket mechanism to mgrlistener to deal with this in some way?
> * the agent code as https connection with verify=false? Is the cephadm listener always going to be self-signed, or should enabling this feature allow for a user provided cert
> * we're sending data to the mgr, from anywhere - you're using a keyring to validate that only our hosts are allowed to send data to mgr..right?
> * the raw socket handling is ipv4 only - is the goal to keep this at ipv4 or support dual stack?
> * any thoughts on how this can be extended to be an agent beyond what the exporter already provides? For example, tell the agent to deploy a daemon, so the mgr offloads not just the data gathering but also the deployment work? As it stands the code is heavily focused on sending state, not enabling the offload of actions to help with scale...right?
> * how are agents going offline being handled? For example, if 1 agent is not sending data is that a healthcheck...what about 10 hosts?

- Since the only thing that needs to happen on mgr failover currently is changing which ip the agents send the metadata to and that ip is being stored in a config file for the agents, I just have it setup to reconfig any agent that was deployed given a target ip that differs from the ip of the current mgr. The _check_agents_ip_port  function in agent.py is responsible for that and it is called as part of the serve loop.
- I sort of want to keep the agent opinionated when possible to keep it simpler. Maybe in the future allowing users to provide their own cert for SSL here would be possible (if anybody wanted it) but I'd like to keep something like that out of the initial implementation as it's adds complexity and another thing that needs to be tested and there's already plenty of that here now.
- Yes, the agent sends it's keyring (which the mgr initially provided it) as part of the POST request along with the metadata. We check that keyring against the one we have stored for that agent and make sure it matches before we'll accept the data. This is done within the RequestHandler class in agent.py.
- I don't see any reason we couldn't do ipv6 support in the future but similar to the user defined SSL cert stuff it's something I didn't want to include here for the sake of keeping this initial implementation simple. There's already a lot that could go wrong here so my idea is to basically get only the most simple working version in in this PR, give it some time to see if errors are found with this basic version, then start adding new functionality when this version feels like it's stable.
- There have already been some discussions on giving the agent this kind of ability and I think it's likely we'll at least try to implement it in the future as it does help with scaling. This current version at least helps with the long repetitive task of gethering metadata on each host that has to be done every time through the serve loop. Deploying and removing daemons are typically done in batches but aren't a continuously happening activity through the lifecycle of the cluster so they're lower priorirt (but not zero priority, like I said we have discussed this and will likely look into it later) and I'm leaving the idea out of this PR. Similar to what I've said on other points, I'd like to keep the initial implementation as simple as possible to try to limit how many points of possible failure we're adding in at once.
- The current way it works, as of me writing this, is that failure of agents happens very quietly. The serve loop is setup to check if we should be using agent metadata or if we want to use the old method of ssh-ing in and grabbing it (_use_agent_metadata function in agent.py). If the agent is up and reporting regularly, this check will fail and the mgr will not ssh in. If we either do not have the setting to use agents on, we have no recorded timestamp for metadata from the agent (covers cases where agent immediately fails after deployment) or if the last timestamp for metadata received from the agent is more than 2.5 x the refresh rate for the agent (covers cases where the agent fails later on) the check will pass, we'll assume we can't expect the agent to provide us with metadata and will fall back to the previous method of ssh-ing in to gather it. If the agent actually is down, this will then be reported in the daemons status (no health warnings currently, like I said it's quite quiet as of now) and, for that host, the mgr will effectively resort back to the old, pre-agent way of doing things. The general idea of this system is that it's safe. The worst case, where the agent just doesn't work, just causes a slight delay and then we're back to using the old methods which we know work (they're slow, but they do work). This limits the damage a bug with this agent can cause which I like because it's a sort of large change and it needs time to be tested more before I'd consider it a stable addition. Now, none of what I said means we couldn't (or shouldn't) also raise a health warning to notify users that an agent or multiple agents are down. I could definitely add that in (and probably should, now that I think about it).

One other note, relating to the last point. The timestamps I'm talking about with the agent do not come from the agent themselves. The mgr generates them when it accepts metadata from an agent. That way all the timestamps come from the same machine so I can be confident they're all consistent. For that same consistency reason, I don't maintain the timestamps across mgr failovers. This means, when the new mgr comes up, there will be no timestamps from any agent and it wil initially assume it needs ssh to gather metadata until the agents have a chance to report. I thought this was okay because 1) we need to reconfig the agents with hte new ip anyway so initially, in a sense, they are in a sort of error state and 2) mgr failovers I expect to be quite infrequent so the performance when they happen is not high priority when compared to normal operation. Once the agents are reconfigured and reporting again it should go back to just using their metadata and not ssh-ing and we'll see the performance improvement in the serve loop again.

Hopefully that answered what you were looking for @pcuzner . This is a fairly large PR so any questions and reviews are appreciated.

I think will look into raising a health check for the agents being down within this PR as well.

yuriw commented:
@adk3798 couldnot merge this PR

From https://github.com/adk3798/ceph
 * branch                    agent1     -> FETCH_HEAD
Auto-merging src/pybind/mgr/orchestrator/_interface.py
Auto-merging src/pybind/mgr/cephadm/tests/test_cephadm.py
Auto-merging src/pybind/mgr/cephadm/tests/fixtures.py
CONFLICT (content): Merge conflict in src/pybind/mgr/cephadm/tests/fixtures.py
Auto-merging src/pybind/mgr/cephadm/serve.py
CONFLICT (content): Merge conflict in src/pybind/mgr/cephadm/serve.py
Auto-merging src/pybind/mgr/cephadm/module.py
Auto-merging src/cephadm/cephadm
Auto-merging debian/control
Auto-merging ceph.spec.in
Automatic merge failed; fix conflicts and then commit the result.

pcuzner commented:
@adk3798 thanks. Just curious - why are the request handlers explicitly defined as their own threads, instead of using cherrypy's threadpool?

adk3798 commented:
> @adk3798 thanks. Just curious - why are the request handlers explicitly defined as their own threads, instead of using cherrypy's threadpool?

Honestly, I just though it looked more organized that way. I liked having a class made to handle the incoming metadata rather than just having the POST function in the cherrypy class be really long. Do you think it would cause a performance hit to have it like that @pcuzner? If so, I can put all the RequestHandler's code back in the POST function.

On a related note, do you think I should be upping the size of the cherrypy threadpool? I think it defaults to 10.

liewegas commented on src/cephadm/cephadm:
```diff
@@ -3439,6 +3453,186 @@ class CephContainer:
+            data = {'host': get_hostname(),
+                    'ls': ls_str,
+                    'networks': networks_str,
+                    'facts': facts_str,
```
why encode ls, network, and facts as json strings?

pcuzner replied:
+1

adk3798 replied:
Done for ls and networks. I left host facts in a string format because as far as I can see, the main way to consume it is through the HostFacts dump function.

liewegas commented on src/pybind/mgr/cephadm/agent.py:
```diff
@@ -0,0 +1,218 @@
+                        up_to_date = True
+                    elif self.host in self.mgr.cache.agent_counter:
+                        # we got old counter value with message, inform agent of new timestamp
+                        self.mgr.agent_helpers._request_agent_acks({self.host})
```
what should happen here in the case where self.host isn't in agent_counter? seems cleaner to have a ``if self.host not in self.mgr.cache.agent_counter`` check at the top and make this block the ``else:`` to avoid duplicating the check

liewegas commented on src/cephadm/cephadm:
```diff
@@ -3439,6 +3453,186 @@ class CephContainer:
+
+    def run(self) -> None:
+        listenSocket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        listenSocket.bind(('0.0.0.0', self.agent.listener_port))
```
Maybe we could specify a starting port (~8000, like ceph daemons?) and try bind in a loop until we find a free one.  Then the port the first time we report in to the mgr.  Otherwise agents from multiple clusters would collide with each other (and/or some other random competing service).

pcuzner replied:
how is the host firewall configured for this? would we need to specify a range of potential ports where the cephadm endpoint binds? What happens at mgr failover time, could the port change forcing a reconfig of all cephadm agents?

adk3798 replied:
Now setup to check the next 1000 ports following the port specified in the mgr config option. Might need to look into firewall stuff, but a lot of our other daemons are doing similar.

liewegas commented on src/pybind/mgr/cephadm/agent.py:
```diff
@@ -0,0 +1,218 @@
+        self.cert_tmp.flush()  # cert_tmp must not be gc'ed
+        cert_fname = self.cert_tmp.name
+
+        verify_tls_files(cert_fname, key_fname)
```
We could additionally provide the cert to the client so that it does a client-side verification too

pcuzner replied:
+1 we shouldn't be using verify=false. The cert build will need to be more generic though (IIRC, I did something similar in the exporter code)

adk3798 replied:
Done. We now generate a root cert and a cert signed by that to have the cephadm endpoint use for ssl. The root cert is given to the agents when deployed in order to verify the ssl cert the cephadm endpoint is using.

liewegas commented:
Overall I like the shape of this!  A few thoughts:
- I think using python3-requests for the agent is a problem.  I'm hoping urllib.request will suffice?  Otherwise we're requiring more than the python standard library...
- Let's add a mode/switch that avoids any of the ssh fallback and relies entirely on the agent?  If the agent doesn't report in, then we try to redeploy it, and if that fails, we mark the host offline?  The idea of falling back feels good in theory but I think we should see if we can make this work reliably/robustly without it.
- The agent listener weirded me out at first but I like it better than a (presumably much slower) ssh call to poke at the agent.

liewegas commented on src/pybind/mgr/cephadm/services/cephadmservice.py:
```diff
@@ -989,3 +990,38 @@ def prepare_create(self, daemon_spec: CephadmDaemonDeploySpec) -> CephadmDaemonD
+        self.mgr.cache.agent_given_ip_port[daemon_spec.host] = (
+            self.mgr.get_mgr_ip(), self.mgr.endpoint_port)
+
+        return config, []
```
That second return list can describe what teh config dependency was.  in our case, I would make it the current mgr ip + port.  Then the existing mgr reconfig machinery will handle redeploying agents after a mgr failover...

liewegas commented on src/pybind/mgr/cephadm/agent.py:
```diff
@@ -0,0 +1,218 @@
+            return False
+        return True
+
+    def _check_agents_ip_port(self) -> None:
```
This can be dropped entirely if you return the mgr ip + port in the config deps from generate_config.

pcuzner commented on src/pybind/mgr/cephadm/agent.py:
```diff
@@ -0,0 +1,218 @@
+        super(CherryPyThread, self).__init__(target=self.run)
+
+    def run(self) -> None:
+        class Root(object):
```
why not move this Root out of the run method and use it to handle the logic instead of starting new threads from a thread that cherrypy has already dispatched? Nitpick - you don't need (object) on the class definition in py3.

something like https://gist.github.com/pcuzner/429135463a88c118ac43b971c84daceb

You could then potentially expose a setting to tweak the cherrypy thread_pool to keep things simple. The default of 10 depends on your service time. In the exporter code a scrape took -18-25ms - so that would be a lot of scrapes in 10s. Whether the default is enough for this depends on your timings - I suspect they'll be fine, but worth checking you push latency to arrive at a sensible default for 200 hosts as a start point.

adk3798 replied:
Mostly done, have a setup similar to your example code now. Can you elaborate somewhat on what you mean by exposing the cherrypy thread pool? Do you think there should be a config option to expand it? Should we grow it ourselves as the number of hosts increases?

pcuzner commented on src/cephadm/cephadm:
```diff
@@ -3439,6 +3453,186 @@ class CephContainer:
+            url = f'https://{self.target_ip}:{self.target_port}'
+            try:
+                requests.post(url, data=data, verify=False)
+            except Exception as e:
```
Is there anyway we can avoid a generic exception here. Maybe there are some specific expcetions we could catch to log different diagnostic messages to the agent log for example?

pcuzner commented on src/cephadm/cephadm:
```diff
@@ -3439,6 +3453,186 @@ class CephContainer:
+            networks = list_networks(self.ctx)
+            networks_str = json.dumps(networks)
+
+            data = {'host': get_hostname(),
```
what happened to the output from ceph-volume? The exporter sent that to pick up on new disks etc, should the agent also be doing this?

adk3798 replied:
added ceph volume output to metadata getting sent by agent

pcuzner commented on src/cephadm/cephadm:
```diff
@@ -3439,6 +3453,186 @@ class CephContainer:
+        super(MgrListener, self).__init__(target=self.run)
+
+    def run(self) -> None:
+        listenSocket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
```
I'm not sure about a bare socket here. It's not clear where this approach would go as and when the agent functionality extends - assuming the goal is to offload as much work as possible to the hosts to achieve scale.

As it stands you can spam this port, and render the data sent to the endpoint unusable - the data is also in the clear. This is the kind of issue that drove the exporter to adopt https after feedback from the rest of the team.

adk3798 replied:
Added 2 way ssl for this socket. Also changed it so it can take on a variable length json string rather than an integer, allowing us to essentially send whatever we want to the agent to make it easy to extend the functionality in the future.

pcuzner commented on src/cephadm/cephadm:
```diff
@@ -3439,6 +3453,186 @@ class CephContainer:
+
+            url = f'https://{self.target_ip}:{self.target_port}'
+            try:
+                requests.post(url, data=data, verify=False)
```
personally I'd be setting the header to json here to make the data easier to deal with at the cherrypy end

pcuzner commented on src/cephadm/cephadm:
```diff
@@ -3439,6 +3453,186 @@ class CephContainer:
+            try:
+                requests.post(url, data=data, verify=False)
+            except Exception as e:
+                logger.debug(f'Failed to send metadata to mgr: {e}')
```
change this to an error level, so it shows in the log.?

pcuzner commented on src/cephadm/cephadm:
```diff
@@ -3439,6 +3453,186 @@ class CephContainer:
+                    'networks': networks_str,
+                    'facts': facts_str,
+                    'ack': str(ack),
+                    'keyring': self.keyring}
```
do you think you should provide the scrape time in the payload sent to the endpoint? If you do that, you could check that against the expected refresh interval to potentially warn about agents falling behind?

adk3798 replied:
It's already generating a timestamp on the mgr side when the metadata is received (using it to tell if the agent is "offline" which is currently that it hasn't reported in 2.5x its refresh rate). Do you think we can just stick with timestamps in the mgr or do we need one from the agent itself?

pcuzner reviewed (changes requested):
Minor nitpicks and suggestions. 

I discussed the agent idea with @sebastian-philipp at the time the exporter code was reviewed, and people we against it, which is why the daemon ended up constained to an exporter feature. So it's good to see that we're back moving in this general direction Once merged, I assume you'll be removing the exporter code, since it's client side functionality is nearly the same as this agent.

yuriw commented:
@adk3798 pls rebase


--- pr 42384 --- pulling https://github.com/adk3798/ceph.git branch agent1
remote: Enumerating objects: 29, done.
remote: Counting objects: 100% (22/22), done.
remote: Total 29 (delta 22), reused 22 (delta 22), pack-reused 7
Unpacking objects: 100% (29/29), 86.19 KiB | 1.57 MiB/s, done.
From https://github.com/adk3798/ceph
 * branch                    agent1     -> FETCH_HEAD
Auto-merging src/pybind/mgr/cephadm/tests/test_cephadm.py
Auto-merging src/pybind/mgr/cephadm/tests/fixtures.py
CONFLICT (content): Merge conflict in src/pybind/mgr/cephadm/tests/fixtures.py
Auto-merging src/pybind/mgr/cephadm/serve.py
CONFLICT (content): Merge conflict in src/pybind/mgr/cephadm/serve.py
Auto-merging src/pybind/mgr/cephadm/module.py
Auto-merging src/cephadm/cephadm
Auto-merging ceph.spec.in
Automatic merge failed; fix conflicts and then commit the result.
Traceback (most recent call last):
  File "/root/wip_master/src/script/build-integration-branch", line 105, in <module>
    assert not r
AssertionError

adk3798 commented:
Update

Changes:

- uses urllib instead of requests library for sending post requests to mgr
- does not have ssh fallback for metadata when use_agent setting is on. If the agent is not reporting for whatever reason the host is marked offline and a CEPHADM_AGENT_DOWN health warning is set
- removed RequestHandler object. Requests are now handled directly in cherrypy POST function (possible followup work related to size of cherrypy thread pool).
- we now generate a root cert and followup cert signed by that root cert for the cephadm endpoint. The endpoint uses the cert signed by the root cert for ssl and the root cert is passed to agents when deploying them so they can verify the ssl cert the endpoint is using. I did also look into 2-way authentication but it doesn't seem to be present in cherrypy. However, between the agent verifying the endpoint's ssl cert and the endpoint verifying the agent keyring both sides are being validated.
- now check 100 ports starting at port specified in a mgr/cephadm config option rather than just giving up if that particular port is taken. (Possible followup related to Firewall)
- Added new function to fixtures that can be used to make the unit tests work with these changes. Originally, once the ssh fallback was removed ~50 tests were failing because no hosts were ever marked as having a daemon refresh


TODO:
- adding ceph-volume output
- reworking MgrListener to be more robust and secure
- clean up around turning "use_agent" config option on/off. Currently doesn't properly remove agents when turning option off and (probably as a result of that issue) hosts are immediately marked offline when turning option back on since last update from agent was past what is considered a "timeout"
- Hosts not properly being marked no longer offline when new metadata from an agent on a host considered offline comes in.

sebastian-philipp commented:
```
h[9314]: cephadm 2021-09-07T05:28:13.658618+0000 mgr.smithi073.qsgvgn (mgr.14162) 724 : cephadm [ERR] [07/Sep/2021:05:28:13] ENGINE Error in HTTPServer.serve
h[9314]: Traceback (most recent call last):
h[9314]:   File "/lib/python3.6/site-packages/cheroot/server.py", line 1810, in serve
h[9314]:     self._connections.run(self.expiration_interval)
h[9314]:   File "/lib/python3.6/site-packages/cheroot/connections.py", line 201, in run
h[9314]:     self._run(expiration_interval)
h[9314]:   File "/lib/python3.6/site-packages/cheroot/connections.py", line 218, in _run
h[9314]:     new_conn = self._from_server_socket(self.server.socket)
h[9314]:   File "/lib/python3.6/site-packages/cheroot/connections.py", line 272, in _from_server_socket
h[9314]:     s, ssl_env = self.server.ssl_adapter.wrap(s)
h[9314]:   File "/lib/python3.6/site-packages/cheroot/ssl/builtin.py", line 278, in wrap
h[9314]:     sock, do_handshake_on_connect=True, server_side=True,
h[9314]:   File "/lib64/python3.6/ssl.py", line 365, in wrap_socket
h[9314]:     _context=self, _session=session)
h[9314]:   File "/lib64/python3.6/ssl.py", line 776, in __init__
h[9314]:     self.do_handshake()
h[9314]:   File "/lib64/python3.6/ssl.py", line 1036, in do_handshake
h[9314]:     self._sslobj.do_handshake()
h[9314]:   File "/lib64/python3.6/ssl.py", line 648, in do_handshake
h[9314]:     self._sslobj.do_handshake()
h[9314]: ssl.SSLError: [SSL: SSLV3_ALERT_BAD_CERTIFICATE] sslv3 alert bad certificate (_ssl.c:897)
``` 

https://pulpito.ceph.com/adking-2021-09-07_04:51:05-rados:cephadm-wip-adk-testing-2021-09-06-1946-distro-basic-smithi/6377523

adk3798 commented:
Most recent run: https://pulpito.ceph.com/adking-2021-09-23_19:49:21-rados:cephadm-wip-adk-testing-2021-09-23-1250-distro-basic-smithi/

1 Failure caused by https://tracker.ceph.com/issues/44587

sebastian-philipp commented:
https://pulpito.ceph.com/adking-2021-09-23_19:49:21-rados:cephadm-wip-adk-testing-2021-09-23-1250-distro-basic-smithi/
